src/backend/distributed/metadata/metadata_utility.c

```diff
@@ -80,11 +85,221 @@ static ShardPlacement * ShardPlacementOnGroup(uint64 shardId, int groupId);
 
 
 /* exports for SQL callable functions */
+PG_FUNCTION_INFO_V1(citus_node_disk_space_stats);
+PG_FUNCTION_INFO_V1(citus_local_disk_space_stats);
+PG_FUNCTION_INFO_V1(citus_database_size);
+PG_FUNCTION_INFO_V1(citus_node_database_size);
 PG_FUNCTION_INFO_V1(citus_table_size);
 PG_FUNCTION_INFO_V1(citus_total_relation_size);
 PG_FUNCTION_INFO_V1(citus_relation_size);
 
 
+/*
+ * citus_node_database_size is a UDF that returns the size of the database
+ * on a given node.
+ */
+Datum
+citus_node_disk_space_stats(PG_FUNCTION_ARGS)
+{
+	text *nodeNameText = PG_GETARG_TEXT_P(0);
+	char *nodeNameString = text_to_cstring(nodeNameText);
+	int32 nodePort = PG_GETARG_INT32(1);
+
+	CheckCitusVersion(ERROR);
+
+	uint64 availableBytes = 0;
+	uint64 totalBytes = 0;
+
+	GetNodeDiskSpaceStats(nodeNameString, nodePort, &availableBytes, &totalBytes);
+
+	TupleDesc tupleDescriptor = NULL;
```

Comment:
Can you put the following into a function so that this function doesn't do things at different levels

```c
static HeapTuple GenerateNodeDiskSpaceTuple(TupleDesc tupleDescriptor, uint64 availableBytes, uint64 totalBytes) {
	Datum values[TABLE_METADATA_FIELDS];
	bool isNulls[TABLE_METADATA_FIELDS];

	memset(values, 0, sizeof(values));
	memset(isNulls, false, sizeof(isNulls));

	values[0] = UInt64GetDatum(availableBytes);
	values[1] = UInt64GetDatum(totalBytes);

	HeapTuple diskSpaceTuple = heap_form_tuple(tupleDescriptor, values, isNulls);
    return distSpaceTuple;
}
```


```c
	TypeFuncClass resultTypeClass = get_call_result_type(fcinfo, NULL,
														 &tupleDescriptor);
	if (resultTypeClass != TYPEFUNC_COMPOSITE)
	{
		ereport(ERROR, (errmsg("return type must be a row type")));
    }
    HeapTuple diskSpaceTuple =GenerateNodeDiskSpaceTuple(tupleDescriptor, availableBytes, totalBytes);
    PG_RETURN_UINT64(HeapTupleGetDatum(diskSpaceTuple));

```


Also it seems that 

```c
	if (get_call_result_type(fcinfo, NULL, &tupleDescriptor) != TYPEFUNC_COMPOSITE)
	{
		elog(ERROR, "return type must be a row type");
	}
```

this is used in `node_protocol.c` and also `backend_data.c`, now we are adding one more usage, we can introduce EnsureFunctionReturnIsRowType(fcinfo, tupleDescriptor) utility function and use it everywhere.

Reply:
> GenerateNodeDiskSpaceTuple

makes sense

> EnsureFunctionReturnIsRowType

I don't like burying fcinfo usage.